### ISSUE TITLE
Dispatch workgroups along X only to prevent exceeding max number of block on GPU

### ISSUE THREAD

**Issue by ThomasRaoux** [iree-org/iree#12642]
### What happened?

Both CUDA and Vulkan may have a low max number (65535) of workgroups along dimensions Y and Z.
Even though this is not a complete solution, one case that should fix the problem for a while is to only dispatch workgroups along X. This is what current XLA does at the moment and seems to have been working for a while.

In order to do that we can change the pass [TileAndDistributeToWorkgroup](https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp) to decide how many dimensions we can dispatch along. Right now this is hardcoded to 3 [here](https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/TileAndDistributeToWorkgroupsPass.cpp#L209), we should change `kNumMaxParallelDims` to be a pass option and use 1 for both LLVMGPU and Vulkan backend.

**Comment by ThomasRaoux** [comment]
@antiagainst does that make sense to you?

**Comment by benvanik** [comment]
This won't work for vulkan - we need to evenly divide out xyz:
https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeWorkGroupCount[0]&platform=all
(that's a large number of targets with 65k max on x)

**Comment by ThomasRaoux** [comment]
> This won't work for vulkan - we need to evenly divide out xyz: [https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeWorkGroupCount[0]&platform=all](https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeWorkGroupCount%5B0%5D&platform=all) (that's a large number of targets with 65k max on x)

wow this is crazy low. The GPUs with the low limit seem pretty old but I'm guessing we still want to support them. Another solution is to divide along the 3 dimension then linearize and de-linearize again based on limit. This probably starts with distributing along 1 dimension and then building this id by linearizing several dimensions.

**Comment by benvanik** [comment]
Mid- to long-term we really want to be semantically using the workgroup IDs to denote locality - linearizing and dividing out is usually a good starting point and if it helps this issue that'd be awesome. We need that on CPU and it can help GPU too (I'm not sure we're too bad there today, but also don't think anyone has verified for example that when we have increasing x we're actually traversing memory in increasing x).

**Comment by qedawkins** [comment]
FYI It looks like there's a related issue in Vulkan with this IR: https://storage.googleapis.com/shark-public/prashant/quant/golden_cpu.mlir

```
[VULKAN] ! Validation Error: [ VUID-vkCmdDispatch-groupCountY-00387 ] Object 0: handle = 0x55cc14c88590, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0xfb655004 | vkCmdDispatch(): groupCountY (102400) exceeds device limit maxComputeWorkGroupCount[1] (65535). The Vulkan spec states: groupCountY must be less than or equal to VkPhysicalDeviceLimits::maxComputeWorkGroupCount[1] (https://vulkan.lunarg.com/doc/view/1.3.239.0/linux/1.3-extensions/vkspec.html#VUID-vkCmdDispatch-groupCountY-00387)
```

(cc @pashu123 to confirm the repro) with these compile commands. Needs some work to isolate the dispatches causing this though.
```
iree-compile --iree-input-type=none --iree-hal-target-backends=vulkan -iree-vulkan-target-triple=ampere-unknown-linux --iree-stream-resource-index-bits=64 --iree-vm-target-index-bits=64 --iree-preprocessing-pass-pipeline="builtin.module(iree-flow-detach-elementwise-from-named-ops,iree-flow-convert-1x1-filter-conv2d-to-matmul,iree-preprocessing-convert-conv2d-to-img2col,iree-preprocessing-pad-linalg-ops{pad-size=16})"  golden_cpu.mlir -o nquant.vmfb
```
```
iree-run-module --device=vulkan --function=forward --input=2x4x64x64xf32=0.5 --input=xi64=2.0 --input=2x77x1024xf32=0.5   --module=nquant.vmfb --vulkan_debug_utils=true --vulkan_debug_verbosity=4 --vulkan_validation_layers=true
```

**Comment by benvanik** [comment]
Another thing we're going to need to handle is collapsing dimensions and blocking - there's no GPU with 20 million threads, so scheduling 20 million threads is silly - we should oversubscribe the device but not by several orders of magnitude :)
We should be able to generate workgroups (thread blocks, etc) that do more than one thing and use that to handle these overflows.

**Comment by powderluv** [comment]
Should we split the Vulkan issue into a separate issue ?

**Comment by ThomasRaoux** [comment]
> Mid- to long-term we really want to be semantically using the workgroup IDs to denote locality - linearizing and dividing out is usually a good starting point and if it helps this issue that'd be awesome. We need that on CPU and it can help GPU too (I'm not sure we're too bad there today, but also don't think anyone has verified for example that when we have increasing x we're actually traversing memory in increasing x).

IDs are meant to represent the order of processing of data. This is already what we do right now and this happens when we do tile and distribute and hopefully we keep it separated from the IDs.



> Another thing we're going to need to handle is collapsing dimensions and blocking - there's no GPU with 20 million threads, so scheduling 20 million threads is silly - we should oversubscribe the device but not by several orders of magnitude :)
> We should be able to generate workgroups (thread blocks, etc) that do more than one thing and use that to handle these overflows.

Right, the difference is that it will have a performance cost so we have to decide when we want to use it.

> Should we split the Vulkan issue into a separate issue ?

Ideally we come up with one solution that solves both. Vulkan might be more restricted in which case we should solve for Vulkan and it would solve cuda.

**Comment by benvanik** [comment]
(we'll also need this on CPU too, though generally we put so much more in a workgroup that limits aren't often the issue even though we have them (16-bit z))

**Comment by antiagainst** [comment]
Ah, yeah, when deducing tiling configurations for workgroup (X, Y, Z), we aren't factoring in the # workgroups limit per dimension right now; so it generates such validation errors. We should consider that and adjust the tiling scheme accordingly for a target. +1 to the above linearizing solution.

**Comment by powderluv** [comment]
Will that be part of work @KoolJBlack is doing ?

**Comment by KoolJBlack** [comment]
I'm looking at the tiling scheme for the CUDA backend with reproducer models. Thomas and I will discuss further tomorrow and I'll have more details. 

I haven't worked on the Vulkan backend so I cannot say if its the exact same tiling scheme yet. If the limits are lower, however I can simulate that on CUDA.

**Comment by KoolJBlack** [comment]
I'm working on modifying our `TileAndDistributeToWorkgroup` pass to support dispatching all workgroups along X. As thomas highlighted above, its currently hardcoded dispatch along [X,Y,Z] and there are several dependencies on this arrangement in the compiler that need to be fixed in order to change this. Digging into them atm. 

On my side,I see this error specifically when we encounter a dispatch with more than 3 dimensions, say a 4D `linalg.generic`, we'll collapse anything above 3 down to the 3rd dimension, causing the Z workgroup count to balloon. We can easily exceed the Z workgroup count limit on CUDA this way. Example gist output showing a crash from `bert_large_uncased_tf`: https://gist.github.com/KoolJBlack/428f32008c3760d874a7b99991e9f7ff

After chatting with Thomas, I believe Vulkan is using the same code, so if this works for CUDA the fix should apply there as well.

**Comment by KoolJBlack** [comment]
Update:  #12774 enabled dispatching along x for CUDA. I've confirmed this solves crahses we'd seen earlier on models with large batch sizes. 

So far the main focus of this issue raised by Thomas has been addressed for CUDA. I gather this solution is insufficient for vulkan based on Ben's comment. The suggestion of evenly using X,Y,Z will need further design/iteration (https://github.com/openxla/iree/issues/12642#issuecomment-1470295528). The updates to `TileAndDistributeToWorkgroupsPass` are a good starting point to make this possible, as we're already remapping a single X dimension to the appropriate loops using `affine.map`. Vulkan could do the same in reverse to collapse the X,Y,Z input into a new X (single dispatch) and then reuse the expansion to the appropriate loop sizes. 

@antiagainst leaving up to you where to take this from here? I'll reassign to you or close the issue.
